File: duckdata/pipelink/crosslink/r/shared_memory/cpp_wrapper.R
#' C++ bindings wrapper for CrossLink
#' 
#' This file provides an interface to the C++ implementation of CrossLink
#' for better cross-language interoperability and performance.

# Global variables to track C++ binding availability
.crosslink_cpp_available <- FALSE
.crosslink_cpp_bindings <- NULL

#' Check if C++ bindings are available
#' @return TRUE if C++ bindings are available
is_cpp_available <- function() {
  return(.crosslink_cpp_available)
}

#' Try to load C++ bindings from various locations
#' @return TRUE if bindings were successfully loaded
try_load_cpp_bindings <- function() {
  if (isTRUE(.crosslink_cpp_available)) {
    return(TRUE)
  }
  
  tryCatch({
    # First approach: try Rcpp::sourceCpp with the binding file
    package_dir <- system.file(package = "crosslink")
    project_dir <- dirname(dirname(package_dir))
    
    # Look for the compiled library in possible locations
    lib_paths <- c(
      file.path(project_dir, "r", "libs", "crosslink_r"),
      file.path(project_dir, "cpp", "build", "crosslink_r"),
      file.path(project_dir, "libs", "crosslink_r")
    )
    
    # Add platform-specific extensions
    if (.Platform$OS.type == "windows") {
      lib_paths <- paste0(lib_paths, c(".dll"))
    } else if (Sys.info()["sysname"] == "Darwin") {
      lib_paths <- paste0(lib_paths, c(".so", ".dylib"))
    } else {
      lib_paths <- paste0(lib_paths, c(".so"))
    }
    
    # Try each path
    for (lib_path in lib_paths) {
      if (file.exists(lib_path)) {
        # Use dyn.load to load the library
        dyn.load(lib_path)
        
        # Check if the functions are now available
        if (exists("crosslink_connect", mode = "function") &&
            exists("crosslink_push", mode = "function") &&
            exists("crosslink_pull", mode = "function")) {
          .crosslink_cpp_available <<- TRUE
          return(TRUE)
        }
      }
    }
    
    # Second approach: try Rcpp::sourceCpp with the binding file directly
    cpp_file <- file.path(project_dir, "cpp", "bindings", "r_binding.cpp")
    if (file.exists(cpp_file)) {
      Rcpp::sourceCpp(cpp_file)
      .crosslink_cpp_available <<- TRUE
      return(TRUE)
    }
    
    # If we get here, we failed to load the bindings
    message("C++ bindings not found in any expected location")
    return(FALSE)
  }, error = function(e) {
    message("Error loading C++ bindings: ", e$message)
    return(FALSE)
  })
}

# Try to load C++ bindings on package load
.onLoad <- function(libname, pkgname) {
  try_load_cpp_bindings()
}

#' CrossLink C++ connection
#' @export
#' @param db_path Path to the DuckDB database file
#' @param debug Enable debug mode with verbose logging
#' @param force_native Force using native R implementation even if C++ is available
#' @return A CrossLink C++ connection object
crosslink_cpp_connect <- function(db_path = "crosslink.duckdb", debug = FALSE, force_native = FALSE) {
  if (force_native || !try_load_cpp_bindings()) {
    if (debug) message("C++ bindings are not available, falling back to native R implementation.")
    return(NULL)
  }
  
  tryCatch({
    cl_ptr <- crosslink_connect(db_path, debug)
    
    # Create a connection object
    cl <- list(
      ptr = cl_ptr,
      db_path = db_path,
      debug = debug
    )
    
    class(cl) <- "crosslink_cpp_connection"
    return(cl)
  }, error = function(e) {
    if (debug) message("Error connecting via C++ bindings: ", e$message, ", falling back to native R implementation.")
    return(NULL)
  })
}

#' Push a data frame or Arrow table to CrossLink via C++ bindings
#' @export
#' @param cl CrossLink C++ connection
#' @param data Data frame or Arrow table to push
#' @param name Optional name for the dataset
#' @param description Optional description
#' @return Dataset ID or NULL if operation failed
push_via_cpp <- function(cl, data, name = "", description = "") {
  if (!inherits(cl, "crosslink_cpp_connection")) {
    if (is.list(cl) && cl$debug) message("Not a valid CrossLink C++ connection, falling back to native R implementation.")
    return(NULL)
  }
  
  tryCatch({
    # Convert data to Arrow table if needed
    if (!inherits(data, "Table")) {
      if (!requireNamespace("arrow", quietly = TRUE)) {
        if (cl$debug) message("The arrow package is required for C++ data sharing")
        return(NULL)
      }
      data <- arrow::as_arrow_table(data)
    }
    
    # Push data using C++ bindings
    return(crosslink_push(cl$ptr, data, name, description))
  }, error = function(e) {
    if (cl$debug) message("Error in C++ push operation: ", e$message, ", falling back to native R implementation.")
    return(NULL)
  })
}

#' Pull data from CrossLink via C++ bindings
#' @export
#' @param cl CrossLink C++ connection
#' @param identifier Dataset ID or name
#' @param to_data_frame Convert result to data frame (default TRUE)
#' @return Arrow table or data frame, or NULL if operation failed
pull_via_cpp <- function(cl, identifier, to_data_frame = TRUE) {
  if (!inherits(cl, "crosslink_cpp_connection")) {
    if (is.list(cl) && cl$debug) message("Not a valid CrossLink C++ connection, falling back to native R implementation.")
    return(NULL)
  }
  
  tryCatch({
    # Pull data using C++ bindings
    arrow_table <- crosslink_pull(cl$ptr, identifier)
    
    # Convert to data frame if requested
    if (to_data_frame) {
      if (!requireNamespace("arrow", quietly = TRUE)) {
        if (cl$debug) message("The arrow package is required for C++ data sharing")
        return(NULL)
      }
      return(arrow::as.data.frame(arrow_table))
    } else {
      return(arrow_table)
    }
  }, error = function(e) {
    if (cl$debug) message("Error in C++ pull operation: ", e$message, ", falling back to native R implementation.")
    return(NULL)
  })
}

#' Query CrossLink database via C++ bindings
#' @export
#' @param cl CrossLink C++ connection
#' @param sql SQL query
#' @param to_data_frame Convert result to data frame (default TRUE)
#' @return Arrow table or data frame with query results, or NULL if operation failed
query_via_cpp <- function(cl, sql, to_data_frame = TRUE) {
  if (!inherits(cl, "crosslink_cpp_connection")) {
    if (is.list(cl) && cl$debug) message("Not a valid CrossLink C++ connection, falling back to native R implementation.")
    return(NULL)
  }
  
  tryCatch({
    # Execute query using C++ bindings
    arrow_table <- crosslink_query(cl$ptr, sql)
    
    # Convert to data frame if requested
    if (to_data_frame) {
      if (!requireNamespace("arrow", quietly = TRUE)) {
        if (cl$debug) message("The arrow package is required for C++ data sharing")
        return(NULL)
      }
      return(arrow::as.data.frame(arrow_table))
    } else {
      return(arrow_table)
    }
  }, error = function(e) {
    if (cl$debug) message("Error in C++ query operation: ", e$message, ", falling back to native R implementation.")
    return(NULL)
  })
}

#' List datasets via C++ bindings
#' @export
#' @param cl CrossLink C++ connection
#' @return Vector of dataset IDs or NULL if operation failed
list_datasets_via_cpp <- function(cl) {
  if (!inherits(cl, "crosslink_cpp_connection")) {
    if (is.list(cl) && cl$debug) message("Not a valid CrossLink C++ connection, falling back to native R implementation.")
    return(NULL)
  }
  
  tryCatch({
    # List datasets using C++ bindings
    return(crosslink_list_datasets(cl$ptr))
  }, error = function(e) {
    if (cl$debug) message("Error in C++ list_datasets operation: ", e$message, ", falling back to native R implementation.")
    return(NULL)
  })
}

#' Register notification callback via C++ bindings
#' @export
#' @param cl CrossLink C++ connection
#' @param callback Function to call when data changes
#' @return Registration ID or NULL if operation failed
register_notification_via_cpp <- function(cl, callback) {
  if (!inherits(cl, "crosslink_cpp_connection")) {
    if (is.list(cl) && cl$debug) message("Not a valid CrossLink C++ connection, falling back to native R implementation.")
    return(NULL)
  }
  
  tryCatch({
    # Register callback using C++ bindings
    return(crosslink_register_notification(cl$ptr, callback))
  }, error = function(e) {
    if (cl$debug) message("Error in C++ register_notification operation: ", e$message, ", falling back to native R implementation.")
    return(NULL)
  })
}

#' Unregister notification callback via C++ bindings
#' @export
#' @param cl CrossLink C++ connection
#' @param registration_id Registration ID from register_notification_via_cpp
#' @return TRUE if successful, FALSE otherwise
unregister_notification_via_cpp <- function(cl, registration_id) {
  if (!inherits(cl, "crosslink_cpp_connection")) {
    if (is.list(cl) && cl$debug) message("Not a valid CrossLink C++ connection, falling back to native R implementation.")
    return(FALSE)
  }
  
  tryCatch({
    # Unregister callback using C++ bindings
    crosslink_unregister_notification(cl$ptr, registration_id)
    return(TRUE)
  }, error = function(e) {
    if (cl$debug) message("Error in C++ unregister_notification operation: ", e$message, ", falling back to native R implementation.")
    return(FALSE)
  })
}

#' Clean up resources via C++ bindings
#' @export
#' @param cl CrossLink C++ connection
#' @return TRUE if successful, FALSE otherwise
cleanup_via_cpp <- function(cl) {
  if (!inherits(cl, "crosslink_cpp_connection")) {
    if (is.list(cl) && cl$debug) message("Not a valid CrossLink C++ connection, falling back to native R implementation.")
    return(FALSE)
  }
  
  tryCatch({
    # Clean up using C++ bindings
    crosslink_cleanup(cl$ptr)
    return(TRUE)
  }, error = function(e) {
    if (cl$debug) message("Error in C++ cleanup operation: ", e$message, ", falling back to native R implementation.")
    return(FALSE)
  })
}

# Try to load C++ bindings
try_load_cpp_bindings() 

File: duckdata/pipelink/crosslink/python/shared_memory/cpp_wrapper.py
"""
C++ binding wrapper for CrossLink Python

This module loads the C++ bindings for CrossLink and provides an interface
to the core C++ implementation for better cross-language zero-copy performance.
"""
import os
import sys
import importlib.util
from pathlib import Path
import warnings

# Define variables for C++ binding availability
_cpp_available = False
_cpp_module = None

# Try to import the C++ extension module
def _try_import_cpp():
    """Try to import the C++ module from various locations"""
    global _cpp_available, _cpp_module
    
    # If already loaded, don't try again
    if _cpp_available:
        return True
    
    try:
        # First method: try direct import if installed in Python path
        import crosslink.crosslink_cpp as cpp_mod
        _cpp_module = cpp_mod
        _cpp_available = True
        return True
    except ImportError:
        pass
    
    # Second method: try to load from local path
    try:
        # Get the potential paths
        current_dir = Path(__file__).parent.resolve()
        project_dir = current_dir.parent.parent
        
        # Look in multiple possible locations
        possible_paths = [
            current_dir / "crosslink_cpp.so",  # Linux/macOS
            current_dir / "crosslink_cpp.pyd",  # Windows
            project_dir / "cpp" / "build" / "crosslink_cpp.so",
            project_dir / "cpp" / "build" / "crosslink_cpp.pyd",
            project_dir / "cpp" / "crosslink_cpp.so",
            project_dir / "cpp" / "crosslink_cpp.pyd"
        ]
        
        for path in possible_paths:
            if path.exists():
                # Load using importlib
                spec = importlib.util.spec_from_file_location("crosslink_cpp", path)
                cpp_mod = importlib.util.module_from_spec(spec)
                spec.loader.exec_module(cpp_mod)
                
                _cpp_module = cpp_mod
                _cpp_available = True
                return True
    except Exception as e:
        warnings.warn(f"Warning: Failed to load C++ module: {e}")
    
    # If we get here, no method succeeded
    warnings.warn("C++ bindings not found in any expected location, falling back to Python implementation")
    return False

# Try to import when module is loaded
_try_import_cpp()

class CrossLinkCpp:
    """Python wrapper for the C++ CrossLink implementation"""
    
    def __init__(self, db_path="crosslink.duckdb", debug=False):
        """Initialize the C++ CrossLink instance
        
        Args:
            db_path: Path to the database file
            debug: Enable debug logging
        
        Raises:
            ImportError: If C++ bindings are not available
        """
        global _cpp_available, _cpp_module
        
        if not _cpp_available:
            if not _try_import_cpp():  # Try one more time
                raise ImportError("CrossLink C++ bindings are not available")
        
        self._cpp_impl = _cpp_module.CrossLinkCpp(db_path, debug)
        self.db_path = db_path
        self.debug = debug
    
    def push(self, table, name="", description=""):
        """Push a PyArrow table to CrossLink
        
        Args:
            table: PyArrow Table
            name: Optional name for the dataset
            description: Optional description
        
        Returns:
            str: Dataset ID
        
        Raises:
            RuntimeError: If push operation fails
        """
        try:
            return self._cpp_impl.push(table, name, description)
        except Exception as e:
            if self.debug:
                print(f"C++ push operation failed: {e}")
            raise RuntimeError(f"Failed to push data: {e}")
    
    def pull(self, identifier):
        """Pull a dataset as a PyArrow table
        
        Args:
            identifier: Dataset ID or name
        
        Returns:
            pyarrow.Table: Arrow table
        
        Raises:
            RuntimeError: If pull operation fails
        """
        try:
            return self._cpp_impl.pull(identifier)
        except Exception as e:
            if self.debug:
                print(f"C++ pull operation failed: {e}")
            raise RuntimeError(f"Failed to pull data for {identifier}: {e}")
    
    def query(self, sql):
        """Execute a SQL query and return results as a PyArrow table
        
        Args:
            sql: SQL query string
        
        Returns:
            pyarrow.Table: Query results
            
        Raises:
            RuntimeError: If query operation fails
        """
        try:
            return self._cpp_impl.query(sql)
        except Exception as e:
            if self.debug:
                print(f"C++ query operation failed: {e}")
            raise RuntimeError(f"Failed to execute query: {e}")
    
    def list_datasets(self):
        """List available datasets
        
        Returns:
            list: List of dataset IDs
            
        Raises:
            RuntimeError: If list operation fails
        """
        try:
            return self._cpp_impl.list_datasets()
        except Exception as e:
            if self.debug:
                print(f"C++ list_datasets operation failed: {e}")
            raise RuntimeError(f"Failed to list datasets: {e}")
    
    def register_notification(self, callback):
        """Register a callback for dataset notifications
        
        Args:
            callback: Function to call when a dataset changes
        
        Returns:
            str: Registration ID
            
        Raises:
            RuntimeError: If registration fails
        """
        try:
            return self._cpp_impl.register_notification(callback)
        except Exception as e:
            if self.debug:
                print(f"C++ register_notification operation failed: {e}")
            raise RuntimeError(f"Failed to register notification: {e}")
    
    def unregister_notification(self, registration_id):
        """Unregister a notification callback
        
        Args:
            registration_id: Registration ID from register_notification
            
        Raises:
            RuntimeError: If unregistration fails
        """
        try:
            return self._cpp_impl.unregister_notification(registration_id)
        except Exception as e:
            if self.debug:
                print(f"C++ unregister_notification operation failed: {e}")
            raise RuntimeError(f"Failed to unregister notification: {e}")
    
    def cleanup(self):
        """Clean up all resources
        
        Raises:
            RuntimeError: If cleanup fails
        """
        try:
            return self._cpp_impl.cleanup()
        except Exception as e:
            if self.debug:
                print(f"C++ cleanup operation failed: {e}")
            raise RuntimeError(f"Failed to cleanup resources: {e}")
    
    def get_from_shared_memory(self, shared_memory_key, metadata=None):
        """Get data from shared memory
        
        Args:
            shared_memory_key: Shared memory key
            metadata: Optional metadata about the shared memory
            
        Returns:
            pyarrow.Table or None: Retrieved data or None if not found
            
        Raises:
            RuntimeError: If operation fails (but not if key not found)
        """
        try:
            # Handle None or empty key
            if shared_memory_key is None or shared_memory_key == "":
                if self.debug:
                    print("Cannot get from shared memory: key is None or empty")
                return None
                
            return self._cpp_impl.get_from_shared_memory(shared_memory_key, metadata or {})
        except KeyError:
            # Specific exception for key not found
            if self.debug:
                print(f"Shared memory key not found: {shared_memory_key}")
            return None
        except Exception as e:
            if self.debug:
                print(f"C++ get_from_shared_memory operation failed: {e}")
            raise RuntimeError(f"Failed to get data from shared memory: {e}")
    
    def setup_shared_memory(self, table, dataset_id):
        """Set up shared memory for a table
        
        Args:
            table: PyArrow Table
            dataset_id: Dataset ID
            
        Returns:
            dict or None: Shared memory information or None if failed
            
        Raises:
            RuntimeError: If setup fails
        """
        try:
            return self._cpp_impl.setup_shared_memory(table, dataset_id)
        except Exception as e:
            if self.debug:
                print(f"C++ setup_shared_memory operation failed: {e}")
            raise RuntimeError(f"Failed to set up shared memory: {e}")
    
    def cleanup_shared_memory(self, shared_memory_key=None):
        """Clean up shared memory
        
        Args:
            shared_memory_key: Optional specific key to clean up, or None for all
            
        Returns:
            int: Number of regions cleaned up
            
        Raises:
            RuntimeError: If cleanup fails
        """
        try:
            if shared_memory_key is not None:
                return self._cpp_impl.cleanup_shared_memory(shared_memory_key)
            else:
                return self._cpp_impl.cleanup_all_shared_memory()
        except Exception as e:
            if self.debug:
                print(f"C++ cleanup_shared_memory operation failed: {e}")
            raise RuntimeError(f"Failed to clean up shared memory: {e}")

# Define module variables for easier importing
is_cpp_available = _cpp_available 